R/arg_reconcile.R

```diff
@@ -0,0 +1,115 @@
+#'                of a compatible type. That object will be returned in the 
+#'                event of error and will have as its \code{"error"} attribute 
+#'                the error object.
+#' @param .finish A function to run on the result before returning it. Ignored 
+#'                if \code{.docall} is \code{FALSE}.
+#'
+#' @return Either a named list or the result of calling \code{fn} with the 
+#'         supplied arguments
+#'
+arg_reconcile <- function(fn, ..., .args = alist(), .docall = FALSE, 
+                          .wtlist = c(), .bklist= c(), .remap = list(), 
+                          .warn = TRUE, .error = "default", .finish = identity) {
+  # capture the formal arguments of the target function
+  frmls <- formals(fn)
+  # both freeform and an explicit list
+  args <- match.call(expand.dots = FALSE)$`...`
+  if (.docall) {
+    for (ii in names(args)) try(args[[ii]] <- eval(args[[ii]], parent.frame()))
+  }
+  # get rid of duplicate arguments, with freeform arguments 
+  dupes <- table(names(args))
+  dupes <- names(dupes[dupes>1])
+  for (ii in dupes) {
+    args[which(names(args) == ii)[-1]] <- NULL
+  }
+  # Merge ... with .args
+  args <- c(args, .args)
+  # Apply whitelist and blacklist. This step also removes duplicates _between_
+  # the freeform (...) and pre-specified (.args) arguments, with ... versions
+  # taking precedence over the .args versions. This is a consequence of the 
+  # intersect() and setdiff() operations and works even if there is no blacklist
+  # nor whitelist
+  if (!missing(.wtlist)) args <- args[intersect(names(args), .wtlist)]
+  args <- args[setdiff(names(args), .bklist)]
+  # if any remappings of one argument to another are specified, perform them
+  for ( ii in names(.remap) ) {
+    if (!.remap[[ii]] %in% names(args) && ii %in% names(args) ) {
+      args[[.remap[[ii]] ]] <- args[[ii]]
+    }
+  }
+  # remove any unnamed arguments
+  args[names(args) == ""] <- NULL
+  # if the target function doesn't have "..." as an argument, check to make sure
+  # only recognized arguments get passed, optionally with a warning
+  if (!"..." %in% names(frmls) ) {
+    unused <- setdiff(names(args), names(frmls))
+    if ( length(unused)>0 ){
+      if (.warn) {
+        warning("The following arguments were ignored for ", 
+                deparse(substitute(fn)), ":\n", paste(unused, collapse = ", "))
+      }
+      args <- args[intersect(names(args), names(frmls))]
+    }
+  }
+  # the final, cleaned-up arguments either get returned as a list or used on the 
+  # function, depending on how .docall is set
+  if ( !.docall ) return(args) else {
+    # run the function and return the result case
+    oo <- try(do.call(fn, args),silent = TRUE)
+    if (!methods::is(oo, "try-error")) return(.finish(oo)) else {
+      # construct an informative error... eventually there will be more 
+      # detailed info here
+      errorhint <- paste('\nThis error was generated by: ',
```

Comment:
add spaces after commas on all lines below this

Reply:
Done